projects/Openfire-IPv6/service.nix

```
@@ -0,0 +1,166 @@
+      isSystemUser = true;
+      group = "openfire";
+    };
+    users.groups.openfire = {};
+
+    systemd.services.openfire-server = {
+      path = [pkgs.rsync];
+      description = "Openfire Server Daemon";
+      serviceConfig = lib.mkMerge [
+        {
+          ExecStart = "${cfg.stateDir}/bin/openfire.sh";
+          User = "openfire";
+          Group = "openfire";
+          Restart = "on-failure";
+          WorkingDirectory = cfg.stateDir;
+        }
+        (lib.mkIf (cfg.stateDir == "/var/lib/openfire") {
+          StateDirectory = "openfire";
+        })
+      ];
+      environment.OPENFIRE_HOME = cfg.stateDir;
+      wantedBy = ["multi-user.target"];
+      after = ["network.target"];
+
+      # Files under `OPENFIRE_HOME` require read-write permissions for Openfire
+      # to work correctly, so we can't directly run it from the nix store.
+      #
+      # Instead, we need to copy those files into a directory which has proper
+      # permissions, but we must only do this once, otherwise we risk
+      # ovewriting server state information every time the server is upgraded.
+      #
+      # As such, if `version` already exists, we assume the rest of
+      # the files do as well, and copy nothing.
+      preStart = let
+        # Update Openfire
+        # https://download.igniterealtime.org/openfire/docs/latest/documentation/upgrade-guide.html
+        updateScript = ''
+          tmpDir="/tmp/openfire-backup"
+          oldVersion=$(cat "${cfg.stateDir}/version")
+          newVersion=$(cat "${cfg.dataDir}/version")
+
+          if [ $oldVersion != $newVersion ]; then
+            echo "Attempting to update Openfire from $oldVersion to $newVersion"
+
+            # Back up the Openfire state directory
+            rsync -a "${cfg.stateDir}/" $tmpDir/
+
+            # Clear old state
+            rm -rf "${cfg.stateDir}/*"
+
+            # Install new state
+            rsync -a --chmod=u=rwX,go=rX "${cfg.package}/opt/" "${cfg.stateDir}/"
+
+            # Copy old configuration
+            # TODO: only backup these directories (+version and plugins)?
+            for dir in conf embedded-db enterprise resources/security; do
+              [ -e $tmpDir/$dir ] && rsync -a $tmpDir/$dir ${cfg.stateDir}/;
+            done
+
+            rsync -a $tmpDir/plugins/ ${cfg.stateDir}/ --exclude=admin
+
+            echo "Update complete"
+          fi
+        '';
+      in ''
+        set -e
+
+        # Install package to state directory (initial run)
+        if [ ! -e "${cfg.stateDir}"/version ]; then
+          rsync -a --chmod=u=rwX,go=rX "${cfg.package}/opt/" "${cfg.stateDir}/"
+        else
+          ${
+          if cfg.autoUpdateState
+          then updateScript
+          else ''
+            echo 'The installed Openfire version does not match the local state. Please consider pinning the package or migrating the state (https://download.igniterealtime.org/openfire/docs/latest/documentation/upgrade-guide.html).'
```

I think the instructions can be more precise:

```suggestion
            echo You are trying to run Openfire ${cfg.package.version}
            echo with a systemd state directory created by Openfire "$(cat ${cfg.stateDir}/version)".
            echo This may render the package unusable and destroy your data.
            echo
            echo Possible workarounds:
            echo 1. Pin the Openfire package to "$(cat ${cfg.stateDir}/version)" and continue using your existing data.
            echo 2. Export your data from Openfire "$(cat ${cfg.stateDir}/version)",
            echo     clear the state directory, and
            echo     import your data to Openfire ${cfg.package.version}.
            echo
            echo     See Openfire documentation for migrating state:
            echo     https://download.igniterealtime.org/openfire/docs/latest/documentation/upgrade-guide.html
```

Feel free to use a heredoc instead...